File: csharp/Patterns/TemplateMethod.cs
using System;

namespace DesignPatterns.Patterns;

public static class TemplateMethod
{
    private abstract class CaffeineBeverage
    {
        public void Prepare()
        {
            BoilWater();
            Brew();
            PourInCup();
            AddCondiments();
        }

        private void BoilWater() => Console.WriteLine("Boil water");
        private void PourInCup() => Console.WriteLine("Pour into cup");
        protected abstract void Brew();
        protected abstract void AddCondiments();
    }

    private class Tea : CaffeineBeverage
    {
        protected override void Brew() => Console.WriteLine("Steep tea bag");
        protected override void AddCondiments() => Console.WriteLine("Add lemon");
    }

    private class Coffee : CaffeineBeverage
    {
        protected override void Brew() => Console.WriteLine("Brew coffee grounds");
        protected override void AddCondiments() => Console.WriteLine("Add sugar");
    }

    public static void Run()
    {
        Console.WriteLine("\nTemplate Method");
        new Tea().Prepare();
        new Coffee().Prepare();
    }
}

File: python/patterns/template_method.py
from abc import ABC, abstractmethod


class CaffeineBeverage(ABC):
    def prepare(self) -> None:
        self.boil_water()
        self.brew()
        self.pour_in_cup()
        self.add_condiments()

    def boil_water(self) -> None:
        print("Boil water")

    def pour_in_cup(self) -> None:
        print("Pour into cup")

    @abstractmethod
    def brew(self) -> None:
        raise NotImplementedError

    @abstractmethod
    def add_condiments(self) -> None:
        raise NotImplementedError


class Tea(CaffeineBeverage):
    def brew(self) -> None:
        print("Steep tea bag")

    def add_condiments(self) -> None:
        print("Add lemon")


class Coffee(CaffeineBeverage):
    def brew(self) -> None:
        print("Brew coffee grounds")

    def add_condiments(self) -> None:
        print("Add sugar")


def run() -> None:
    print("\nTemplate Method")
    Tea().prepare()
    Coffee().prepare()
